contracts/verifier.sol

```diff
@@ -0,0 +1,260 @@
+                    s.lowHash = s.medHash;
+                } else {
+                    // and if we're requesting the midStep that we've already requested,
+                    //   there's nothing to do.
+                    // @TODO(shrugs) - should this revert?
+                }
+            }
+
+            s.medStep = step;
+            s.medHash = bytes32(0);
+        }
+        s.lastChallengerMessage = now;
+        NewQuery(sessionId);
+    }
+
+    function respond(uint sessionId, uint step, bytes32 hash)
+        onlyClaimant(sessionId)
+        public
+    {
+        VerificationSession storage s = sessions[sessionId];
+        // Require step to avoid replay problems
+        require(step == s.medStep);
+        if (hash == 0) {
+            // Special "fold" signal
+            // @TODO(shrugs) - is this correct? should the claimant not fold here?
+            challengerConvicted(sessionId);
+            return;
+        }
+
+        // record the claimed hash
+        s.medHash = hash;
+        s.lastClaimantMessage = now;
+
+        // notify watchers
+        NewResponse(sessionId);
+    }
+
+    function performStepVerification(
+        uint sessionId,
+        bytes preValue,
+        bytes postValue,
+        bytes proofs,
+        address claimManager
+    )
+        onlyClaimant(sessionId)
+        public
+    {
+        VerificationSession storage s = sessions[sessionId];
+        require(s.lowStep + 1 == s.highStep);
+        // ^ must be at the end of the binary search according to the smart contract
+
+        if (keccak256(preValue) != s.lowHash) { claimantConvicted(sessionId); return; }
+        if (keccak256(postValue) != s.highHash) { return claimantConvicted(sessionId); return; }
+        // ^ if the claimant supplied anything other than what we know, convict them
+        // @TODO(shrugs) - why do we even have this if we already know this information?
+
+        ClaimManager cm = ClaimManager(claimManager);
+        if (performStepVerificationSpecific(s, s.lowStep, preValue, postValue, proofs)) {
+            cm.claimDecided(sessionId, s.claimant, s.challenger);
+            challengerConvicted(sessionId);
+        } else {
+            cm.claimDecided(sessionId, s.challenger, s.claimant);
+            claimantConvicted(sessionId);
+        }
+    }
+
+    function performStepVerificationSpecific(
+        VerificationSession storage session,
+        uint step,
+        bytes preState,
+        bytes postState,
+        bytes proof
+    )
+        internal
+        returns (bool);
+
+    function isInitiallyValid(VerificationSession storage session)
+        internal
+        returns (bool);
+
+    function timeout(uint sessionId)
+        public
+    {
+        var session = sessions[sessionId];
+        require(session.claimant != 0);
+        if (
+            session.lastChallengerMessage > session.lastClaimantMessage &&
+            now > session.lastClaimantMessage + responseTime
```

Comment:
Should probably be `lastChallengerMessage`

Reply:
Fixed in 188b5e24c5f802d388ef7f22b1b4f4c86e2feb7d